File: archive/v2-csharp-avalonia/src/Spanova.App/AlternativeRow.cs
using Spanova.Core.Model;
using Spanova.Rules;

namespace Spanova.App;

/// <summary>Flat, bindable view of a <see cref="BridgeAlternative"/> + its <see cref="RuleEngineResult"/> for the results table.</summary>
public sealed class AlternativeRow
{
    public AlternativeRow(BridgeAlternative alternative, RuleEngineResult ruleResult)
    {
        SpanCount = alternative.SpanLayout.SpanCount;
        SpanLengthM = Math.Round(alternative.SpanLayout.Spans.Count > 0 ? alternative.SpanLayout.Spans[0].LengthM : 0, 3);
        TotalLengthM = Math.Round(alternative.SpanLayout.TotalLengthM, 3);
        GirderCount = alternative.Girder.Count;
        GirderDepthM = alternative.Girder.DepthM;

        Status = ruleResult.IsFeasible
            ? (ruleResult.Evaluations.Count == 0 ? "Pass (no rules yet)" : "Pass")
            : "Reject";
        FailureReasons = string.Join("; ", ruleResult.FailureReasons);
    }

    public int SpanCount { get; }
    public double SpanLengthM { get; }
    public double TotalLengthM { get; }
    public int GirderCount { get; }
    public double GirderDepthM { get; }
    public string Status { get; }
    public string FailureReasons { get; }
}

File: backend/api/src/main/java/com/spanova/api/alternatives/AlternativeRow.java
package com.spanova.api.alternatives;

import com.spanova.bridgecore.BridgeAlternative;

/**
 * Flat, frontend-friendly view of a {@link BridgeAlternative} for the
 * P03 results table. A presentation DTO, not a bridge-core domain
 * type - {@code bridge-core} stays free of API/JSON-shaping concerns.
 */
public record AlternativeRow(
        String id,
        int spanCount,
        double spanLengthM,
        double totalLengthM,
        int girderCount,
        double girderDepthM,
        boolean feasible) {

    /**
     * @param feasible result of running {@link com.spanova.rules.RuleEngine}
     *        (P04) against this alternative. With no rules approved yet,
     *        this is always {@code true} - the field exists so the wiring
     *        is real, not decorative, ahead of the first real rule.
     */
    public static AlternativeRow from(BridgeAlternative alternative, boolean feasible) {
        var spans = alternative.spanLayout().spans();
        double spanLengthM = spans.isEmpty() ? 0 : spans.get(0).lengthM();

        return new AlternativeRow(
                alternative.id().toString(),
                alternative.spanLayout().spanCount(),
                spanLengthM,
                alternative.spanLayout().totalLengthM(),
                alternative.girder().count(),
                alternative.girder().depthM(),
                feasible);
    }
}
